libraries/logging/stream-logger-file/src/main/java/io/getstream/logging/file/FileStreamLogger.kt

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2014-2022 Stream.io Inc. All rights reserved.
+ *
+ * Licensed under the Stream License;
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    https://github.com/GetStream/stream-chat-android/blob/main/LICENSE
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.getstream.logging.file
+
+import io.getstream.logging.Priority
+import io.getstream.logging.StreamLogger
+import io.getstream.logging.helper.stringify
+import java.io.BufferedWriter
+import java.io.Closeable
+import java.io.File
+import java.io.FileOutputStream
+import java.io.OutputStreamWriter
+import java.io.PrintWriter
+import java.io.Writer
+import java.text.DateFormat
+import java.text.SimpleDateFormat
+import java.util.Date
+import java.util.Locale
+import java.util.concurrent.Executor
+import java.util.concurrent.Executors
+
+private const val DEFAULT_SIZE: Int = 12 * 1024 * 1024
+
+private const val SHAREABLE_FILE: String = "stream_log_%s.txt"
+private const val INTERNAL_FILE_0: String = "internal_0.txt"
+private const val INTERNAL_FILE_1: String = "internal_1.txt"
+
+/**
+ * The [StreamLogger] implementation with log file persistence.
+ */
+public class FileStreamLogger(
+    private val config: Config,
+) : StreamLogger {
+
+    private val executor: Executor = Executors.newSingleThreadExecutor()
+
+    private val timeFormat: DateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss''SSS", Locale.ENGLISH)
+    private val dateFormat: DateFormat = SimpleDateFormat("yyMMddHHmm_ss", Locale.ENGLISH)
+
+    private val internalFile0: File = File(config.filesDir, INTERNAL_FILE_0)
+    private val internalFile1: File = File(config.filesDir, INTERNAL_FILE_1)
+    private val internalFiles: Array<File> = arrayOf(internalFile0, internalFile1)
+
+    private var currentFile: File? = null
+    private var currentWriter: Writer? = null
+
+    init {
+        executor.execute {
+            init()
+        }
+    }
+
```

Comment:
Perform logic, and on this case background tasks, is not a good idea.
As an example, if we would need to mock/spy this class in our tests, we won't be able to avoid this logic, that maybe is not needed for our tests.
Could you considere to extract it outside of the class??
One way could be to have a private constructor that receive all needed arguments, and having a companion method to build this an instance of this class

Reply:
There is a specified implementation for Tests named `KotlinStreamLogger`

Reply:
It is fine that we have another implementation that could be used on tests, but, what about classes that explicitly ask for a `FileStreamLoger`? It can't be mocked without running this logic.
Keeping `init()` block outside of our classes are good practiced

Reply:
`FileStreamLoger` is used at `StreamLogFileProvider` only. That ContentProvider will be included in Debug builds only. There is no need to mock `FileStreamLoger`.